### PR TITLE
Created tags! Support for creating, deleting, and searching; editing will come soon.

### LINKED ISSUE #73: tag creation command

### PR DESCRIPTION
**Pull request by VelvetToroyashi**
I should probably add editing tags.

### CODE REVIEW

File: src/Silk.Core/Commands/Tests/TagCommand.cs
Comment (ghost) [inline src/Silk.Core/Commands/Tests/TagCommand.cs]: Consider generating this instead of hardcoding if possible

File: src/Silk.Core/Commands/Tests/TagCommand.cs
Comment (OoLunar) [inline src/Silk.Core/Commands/Tests/TagCommand.cs]: Instead of reserving words, couldn't you just make the command priority of the get command lower than subcommands?

**Review by ghost — APPROVED** [review]
Flying over it, it looks fine 

Consider adding more comments tho

File: src/Silk.Core/Commands/Tests/TagCommand.cs
Comment (VelvetToroyashi) [reply src/Silk.Core/Commands/Tests/TagCommand.cs]: how so?

Comment (VelvetToroyashi) [reply src/Silk.Core/Commands/Tests/TagCommand.cs]: Reflection, perhaps?

File: src/Silk.Core/Commands/Tests/TagCommand.cs
Comment (VelvetToroyashi) [reply src/Silk.Core/Commands/Tests/TagCommand.cs]: Theoretically, though that could still error out commands on both ends; people that want to actually do something with a tag, and someone who wants to use a specific tag.

Comment (ghost) [reply src/Silk.Core/Commands/Tests/TagCommand.cs]: Even if possible, this would result in a tag with the name of a reserved word to never work.

i.e. tag create with no parameters -> error, tag create with parameters -> tag created

A tag called `create` would never be displayed this way.

Furthermore, it would probably be rather confusing for end users.

File: src/Silk.Core/Commands/Tests/TagCommand.cs
Comment (ghost) [reply src/Silk.Core/Commands/Tests/TagCommand.cs]: Maybe, there's a way how D#+ gets command names so I'm sure you can fill that array in the ctor using whatever D#+ does.

File: src/Silk.Core/Commands/Tests/TagCommand.cs
Comment (VelvetToroyashi) [reply src/Silk.Core/Commands/Tests/TagCommand.cs]: I do agree with Natsumi.

File: src/Silk.Core/Commands/Tests/TagCommand.cs
Comment (OoLunar) [inline src/Silk.Core/Commands/Tests/TagCommand.cs]: What's a lambda?

File: src/Silk.Core/Services/TagService.cs
Comment (OoLunar) [inline src/Silk.Core/Services/TagService.cs:17]: What's a lambda?

File: src/Silk.Core/Services/TagService.cs
Comment (OoLunar) [inline src/Silk.Core/Services/TagService.cs:21]: What's a lambda?

File: src/Silk.Core/Commands/Tests/TagCommand.cs
Comment (VelvetToroyashi) [reply src/Silk.Core/Commands/Tests/TagCommand.cs]: I suppose. Seems like it would be a bit clunky, but I'll think on that one.

File: src/Silk.Core/Commands/Tests/TagCommand.cs
Comment (VelvetToroyashi) [inline src/Silk.Core/Commands/Tests/TagCommand.cs]: I made an oversight here. Second time I've done this too. Used the User Id instead of Guild Id, even though I'm looking directly at inlined variable names.

File: src/Silk.Core/Services/TagService.cs
Comment (OoLunar) [inline src/Silk.Core/Services/TagService.cs]: While I don't have anything against Tuples, I do think this could be done differently. How often is this function used?

File: src/Silk.Core/Services/TagService.cs
Comment (OoLunar) [inline src/Silk.Core/Services/TagService.cs]: While I don't have anything against Tuples, I do think this could be done differently. How often is this function used?

File: src/Silk.Core/Services/TagService.cs
Comment (OoLunar) [inline src/Silk.Core/Services/TagService.cs]: While I don't have anything against Tuples, I do think this could be done differently. How often is this function used?

File: src/Silk.Data/MediatR/Handlers/TagRequestHandler.cs
Comment (OoLunar) [inline src/Silk.Data/MediatR/Handlers/TagRequestHandler.cs:17]: Lambda

File: src/Silk.Data/MediatR/Handlers/TagRequestHandler.cs
Comment (OoLunar) [inline src/Silk.Data/MediatR/Handlers/TagRequestHandler.cs:45]: Lambda

File: src/Silk.Data/MediatR/Handlers/TagRequestHandler.cs
Comment (OoLunar) [inline src/Silk.Data/MediatR/Handlers/TagRequestHandler.cs:66]: Lambda

File: src/Silk.Data/MediatR/Handlers/TagRequestHandler.cs
Comment (OoLunar) [inline src/Silk.Data/MediatR/Handlers/TagRequestHandler.cs:88]: Lambda

File: src/Silk.Data/MediatR/Handlers/TagRequestHandler.cs
Comment (OoLunar) [inline src/Silk.Data/MediatR/Handlers/TagRequestHandler.cs:125]: Lambda

File: src/Silk.Data/MediatR/Handlers/TagRequestHandler.cs
Comment (OoLunar) [inline src/Silk.Data/MediatR/Handlers/TagRequestHandler.cs:146]: Lambda

File: src/Silk.Data/Migrations/20210303085535_Tags_IdIncrement.cs
Comment (OoLunar) [inline src/Silk.Data/Migrations/20210303085535_Tags_IdIncrement.cs:7]: What's this?

**Comment by VelvetToroyashi** [comment]
Everything seems to be working. Waiting on @CalebABG for approval, and then merging to dev.

File: src/Silk.Data/Migrations/20210303183406_Tags_Add_FK.cs
Comment (OoLunar) [inline src/Silk.Data/Migrations/20210303183406_Tags_Add_FK.cs:7]: What's this?

File: src/Silk.Data/ModelConfigurations/TagConfiguration.cs
Comment (OoLunar) [inline src/Silk.Data/ModelConfigurations/TagConfiguration.cs:10]: Lambda

**Review by OoLunar — APPROVED** [review]

File: src/Silk.Data/Migrations/20210303085535_Tags_IdIncrement.cs
Comment (VelvetToroyashi) [reply src/Silk.Data/Migrations/20210303085535_Tags_IdIncrement.cs:7]: Auto-generated by EFCore.

File: src/Silk.Core/Services/TagService.cs
Comment (VelvetToroyashi) [reply src/Silk.Core/Services/TagService.cs]: As often as tags are updated. So, hard to say in practice.

File: src/Silk.Core/Commands/Tests/TagCommand.cs
Comment (VelvetToroyashi) [reply src/Silk.Core/Commands/Tests/TagCommand.cs]: It's done for consistency. I could use an inline constructor, but it'd need to refactored should this class ever require more dependancies.

File: src/Silk.Core/Commands/Tests/TagCommand.cs
Comment (ChaseDRedmon) [inline src/Silk.Core/Commands/Tests/TagCommand.cs]: What does .StartsWith extension method do? I would expect it to take an argument.

Like, `tagname.StartsWith("content");`

 The name just seems weird.
 
 Or should this just be
 
 `if(_reservedWords.Any(tagName))`

Comment (VelvetToroyashi) [reply src/Silk.Core/Commands/Tests/TagCommand.cs]: `_reservedWords` is a collection of strings, and so in C# you can pass the method group in place of a lambda expression, and the variable gets substituted in. Or at least that's my understanding of it.

It's the same as doing `_reservedWords.Any(w => tagName.StartsWith(w));`

File: src/Silk.Core/Commands/Tests/TagCommand.cs
Comment (ProfDoof) [reply src/Silk.Core/Commands/Tests/TagCommand.cs]:
```cs
this.GetTypeInfo().DeclaredMethods.Where(IsCommandCandidate).Select(GetCommandName);

bool IsCommandCandidate(MethodInfo m) 
{
    if (m is null || m.IsStatic || m.IsAbstract || m.IsConstructor || m.IsSpecialName)
        return false;
    var p = m.GetParameters();
    if (!p.Any() || p.First().ParameterType != typeof(CommandContext) || m.ReturnType != typeof(Task)
        return false;
    var a = m.GetCustomAttributes();
    if (a.FirstOrDefault(xa => xa is CommandAttribute) == null)
        return false;
    return true;
}

string GetCommandName(MethodInfo m)
{
    var a = ;
    return m.GetCustomAttributes().FirstOrDefault(xa => xa is CommandAttribute)?.Name ?? m.Name.EndsWith("Async") && m.Name != "Async" ? m.Name.Substring(0, m.Name.Length - 5) : m.Name;
}
```

Here's basically how you would do that based off how D#+ does it. If you are going to do this I would recommend putting them into a thread-safe singleton and passing them in, that way you only need to do the generation once. 
(I only did this as parts of a one-liner cuz I was in GitHub, please don't write this, in reality, this way, also I don't guarantee this working)

File: src/Silk.Core/Commands/Tests/TagCommand.cs
Comment (ChaseDRedmon) [reply src/Silk.Core/Commands/Tests/TagCommand.cs]: yes, but why do you need `.StartsWith`? Doesn't `_reservedWords.Any(tagName)` check the collection of strings to see if the `tagName` is reserved in the collection?

File: src/Silk.Core/Services/TagService.cs
Comment (ProfDoof) [inline src/Silk.Core/Services/TagService.cs]: I've seen you use Success, Reason in multiple places at this point. You might as well just define a short class at this point that does this. Makes it easier to add extra info later if you decide to do so.

File: src/Silk.Core/Commands/Tests/TagCommand.cs
Comment (VelvetToroyashi) [reply src/Silk.Core/Commands/Tests/TagCommand.cs]: I think you're thinking of `.Contains()`, which doesn't exactly fulfill the same purpose. I use `.StartsWith` to prevent tags from interfering with commands. e.g: `!tag create "create a player" To create a player...`, someone that wanted to use that tag could never use it, since one of the tag commands is `!tag create`

Comment (VelvetToroyashi) [reply src/Silk.Core/Commands/Tests/TagCommand.cs]: However, something like `!tag create "how to create a player" ...` would be valid, as the first argument isn't being parsed as a command.

File: src/Silk.Data/MediatR/Handlers/TagRequestHandler.cs
Comment (ProfDoof) [inline src/Silk.Data/MediatR/Handlers/TagRequestHandler.cs:97]: Hold on, is this a foreign key to a tag in the same table?

File: src/Silk.Data/Models/Tag.cs
Comment (ProfDoof) [inline src/Silk.Data/Models/Tag.cs:21]: You diddddddd.... whyyyyyyyy?????? Yes, you need to split this into two tables. One for aliases, one for all the tag info. And you should define the configuration either using the Fluent API or with the Attributes. Take your pick cuz this is nassssstttyyyyy.

**Review by ProfDoof — CHANGES_REQUESTED** [review]

File: src/Silk.Data/MediatR/Handlers/TagRequestHandler.cs
Comment (VelvetToroyashi) [reply src/Silk.Data/MediatR/Handlers/TagRequestHandler.cs:97]: Yes and no(?)

File: src/Silk.Data/Models/Tag.cs
Comment (VelvetToroyashi) [reply src/Silk.Data/Models/Tag.cs:21]: 1. I initially had two tables. It didn't work out. 
2. Tags and aliases share so much information I don't see the point in making another model and table for them

File: src/Silk.Core/Commands/Tests/TagCommand.cs
Comment (ChaseDRedmon) [reply src/Silk.Core/Commands/Tests/TagCommand.cs]: Ah, ok, I get it now

Comment (VelvetToroyashi) [reply src/Silk.Core/Commands/Tests/TagCommand.cs]: Small correction to my previous statement regarding lambdas, thanks to @333fred 
43151662-6AC7-498E-AEA6-1DBD628971BC

Comment (ProfDoof) [reply src/Silk.Core/Commands/Tests/TagCommand.cs]: In C# when you pass in a method name since they are considered first-class citizens, the function gets passed in. The function accepts a string then so you are then able to invoke the function in the Any function to check if the tagName.StartsWith any of the reserved words.

File: src/Silk.Core/Services/TagService.cs
Comment (VelvetToroyashi) [reply src/Silk.Core/Services/TagService.cs]: Will do.

**Review by CalebABG — APPROVED** [review]

File: src/Silk.Data/MediatR/Handlers/TagRequestHandler.cs
Comment (westonpace) [inline src/Silk.Data/MediatR/Handlers/TagRequestHandler.cs:97]: Minor nit here, no need to include `t.OriginalTag`, you don't reference it anywhere.

File: src/Silk.Data/MediatR/Handlers/TagRequestHandler.cs
Comment (westonpace) [inline src/Silk.Data/MediatR/Handlers/TagRequestHandler.cs:107]: And here you don't need `Aliases`.  I don't think you can have an alias of an alias if I read the earlier logic correctly?  If you can then this isn't going to delete your aliases' aliases.

File: src/Silk.Data/MediatR/Handlers/TagRequestHandler.cs
Comment (westonpace) [inline src/Silk.Data/MediatR/Handlers/TagRequestHandler.cs:134]: Dupe

File: src/Silk.Data/MediatR/Handlers/TagRequestHandler.cs
Comment (westonpace) [inline src/Silk.Data/MediatR/Handlers/TagRequestHandler.cs:139]: null over an empty list seems like an odd convention?  More explosive.  But it's not wrong.

File: src/Silk.Data/MediatR/Handlers/TagRequestHandler.cs
Comment (VelvetToroyashi) [reply src/Silk.Data/MediatR/Handlers/TagRequestHandler.cs:107]: Aliases of aliases do exist and from basic testing they seem to be deleted without issue.

File: src/Silk.Data/MediatR/Handlers/TagRequestHandler.cs
Comment (westonpace) [inline src/Silk.Data/MediatR/Handlers/TagRequestHandler.cs:159]: I'm pretty sure `Like` is case insensitive by default.  I suppose it depends on the actual DB used.

File: src/Silk.Data/MediatR/TagRequest.cs
Comment (westonpace) [inline src/Silk.Data/MediatR/TagRequest.cs:9]: Nit, excess space here.  Get some kind of "format-on-save" extension to save yourself headaches.

File: src/Silk.Data/Models/Tag.cs
Comment (westonpace) [inline src/Silk.Data/Models/Tag.cs:23]: So why isn't aliases just a list of strings?  It seems like you'll have to do extra work to make sure tag content is the same across all aliases and you don't do alias of alias, etc.

File: src/Silk.Core/Commands/Server/TagCommand.cs
Comment (westonpace) [inline src/Silk.Core/Commands/Server/TagCommand.cs:48]: This `Uses` count is only going to show the uses of this particular tag and not its aliases.  I'm not sure what it is intended for but one could imagine you would want the `Uses` to be shared across a tag and all of its aliases.

**Review by westonpace — COMMENTED** [review]
Looks good, thanks for sharing.

File: src/Silk.Data/Models/Tag.cs
Comment (VelvetToroyashi) [reply src/Silk.Data/Models/Tag.cs:23]: Moderation reasons; aliases are almost the same as tags, and they're actually fairly easy to update as I can just check for all tags that have a specific FK. It also allows for servers to see who's making excessive aliases, and restricts who can delete aliases, as aliases have owners.

File: src/Silk.Core/Commands/Server/TagCommand.cs
Comment (VelvetToroyashi) [reply src/Silk.Core/Commands/Server/TagCommand.cs:48]: Perhaps. Hadn't really thought about it, but as I said, aliases are tags, technically speaking, so it's kind of the same side of a coin depending on how you look at it.

File: src/Silk.Data/MediatR/Handlers/TagRequestHandler.cs
Comment (VelvetToroyashi) [reply src/Silk.Data/MediatR/Handlers/TagRequestHandler.cs:159]: Postgres is not, by default unfortunately.

File: src/Silk.Data/MediatR/Handlers/TagRequestHandler.cs
Comment (VelvetToroyashi) [reply src/Silk.Data/MediatR/Handlers/TagRequestHandler.cs:139]: Saves from having to iterate over a collection more than necessary. I also like the `is null` syntax.

File: src/Silk.Data/MediatR/TagRequest.cs
Comment (VelvetToroyashi) [reply src/Silk.Data/MediatR/TagRequest.cs:9]: Alas the one of few issues with Rider. I can format tomorrow when I'm back on my machine.

**Pull request merged**
